### PR TITLE
实时链路接入说明

### PR DESCRIPTION
**Pull request by fenglinglingzi**
#### 概述

本次 PR 在 `temporal` 模块中引入了基于 GRU 的时序推理能力，支持加载自定义 PyTorch 模型进行动作识别和状态判断。主要包括：

- 新增 GRUClassifier, TransformerClassifier 模型定义
- 新增 GRUOperator 基类，支持映射表、设备管理和惰性加载
- 新增 CleanOperator 子类，实现逐帧滑动窗口推理

#### 变更内容

+ `app/services/inference/temporal/model/gru.py` 新增 GRU 模型定义
+ `app/services/inference/temporal/model/transformer.py` 新增 Transformer 模型定义
+ `app/services/inference/temporal/operator.py` 新增 GRUOperator 基类
+ `app/services/inference/workflows/clean.py` 新增 CleanOperator 子类
+ `config/inference_config.yaml` 新增清洗阶段算子配置
+ `docs/update/20260713_TEMPORAL_INFERENCE.md` 本次改动说明文档

#### 自测情况

- [x] 通过 tests/ 回归测试
- [x] CleanOperator 可产出相应事件 Events

#### 注意事项
+ 模型文件（如 `gru-final.pt`）未纳入 Git，需手动放置至 `app/data/`

### CODE REVIEW

File: app/services/inference/temporal/operator.py [hunk 1]
```diff
@@ -18,12 +18,16 @@
 from __future__ import annotations
 
 import logging
+import threading
 from abc import ABC, abstractmethod
 from dataclasses import dataclass
 from typing import Any, Dict, List, Tuple
 
+import torch
```
Comment (Jiadezhende) [inline app/services/inference/temporal/operator.py]: 基类里边不要引入重依赖，看看detector是怎么设计的，目前import torch 和 from ...model import GRUClassifier 写在了模块顶层，所有operators都必须引入torch，不方便做单测

File: app/services/inference/temporal/operator.py [hunk 2]
```diff
@@ -91,3 +95,76 @@ def _zip_by_ts(
...
+        self._model_load_lock = threading.Lock()
+        self._device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    def get_object_name(self, class_id: int) -> str:
+        return self.objects.get(class_id, f"object_{class_id}")
+ 
+    def get_action_name(self, class_id: int) -> str:
+        return self.actions.get(class_id, f"action_{class_id}")
+
+    def _ensure_model_loaded(self) -> None:
```
Comment (Jiadezhende) [inline app/services/inference/temporal/operator.py:137]: 模型初次加载时，记得self._model.eval()切为推理态,关掉 dropout

File: app/services/inference/temporal/operator.py [hunk 3]
```diff
@@ -91,3 +95,76 @@ def _zip_by_ts(
...
+                raise
+
+    def infer(self, features: torch.Tensor) -> List[int]:
+        """GRUClassifier 推理：返回每个时间步的预测类别。"""
+        self._ensure_model_loaded()
+        if features.dim() == 2:
+            features = features.unsqueeze(0)
+
+        features = features.to(self._device)
+        logits = self._model(features)
```
Comment (Jiadezhende) [inline app/services/inference/temporal/operator.py]: forward 包进 torch.no_grad()，关自动求导图

File: app/services/inference/workflows/clean.py [hunk 4]
```diff
@@ -96,3 +103,129 @@ def prepare_visualization_data(self, output: FrameDetections) -> RenderSpec:
...
+            name=name,
+            subscribes=subscribes,
+            window_seconds=window_seconds,
+            model_path=model_path,
+            actions=actions,
+            objects=objects,
+            hidden=hidden,
+            num_layers=num_layers,
+        )
+        self.history_frames = []
```
Comment (Jiadezhende) [inline app/services/inference/workflows/clean.py]: CleanOperator 不用自己维护 history_frames 这个帧累加器，_zip_by_ts/_clip 已经按 window_seconds 实现了

File: app/services/inference/workflows/clean.py [hunk 5]
```diff
@@ -96,3 +103,129 @@ def prepare_visualization_data(self, output: FrameDetections) -> RenderSpec:
...
+    def analyze(self, windows: Dict[str, List[FrameDetections]]) -> None:
+        """消费订阅流、推进 self._sm。windows: {流名: 该流滑窗快照(按 ts 升序)}。"""
+        aligned_frames = self._zip_by_ts(windows)
+        if not aligned_frames:
+            return        
+        self._advance(aligned_frames)
+
+    def judge(self) -> Tuple[List[str], List[Alarm]]:
+        """读 self._sm 出 (overlay 文案, 实时告警)。"""
+        events = (
```
Comment (Jiadezhende) [inline app/services/inference/workflows/clean.py]: 多余括号

File: app/services/inference/workflows/clean.py [hunk 6]
```diff
@@ -96,3 +103,129 @@ def prepare_visualization_data(self, output: FrameDetections) -> RenderSpec:
...
+
+            class_features = {}
+            for stream_name, frame_detections in aligned.by_source.items():
+                for detection in frame_detections.detections:
+                    class_features.setdefault(detection.class_id, []).append(detection)
+
+            for class_id, detections in class_features.items():
+                if class_id >= self.num_objects:
+                    continue
+                best_detection = max(detections, key=lambda x: x.confidence)
```
Comment (Jiadezhende) [inline app/services/inference/workflows/clean.py]: 这里的过滤方式不合理，hand目标有两个

File: config/inference_config.yaml
Comment (Jiadezhende) [inline config/inference_config.yaml:1]: yaml此处主要起流程编排的作用，模型结构等信息最好能直接和权重绑定，查下存模型配置的方法

**Review by Jiadezhende — CHANGES_REQUESTED** [review]

File: app/services/inference/workflows/clean.py [hunk 7]
```diff
@@ -96,3 +102,111 @@ def prepare_visualization_data(self, output: FrameDetections) -> RenderSpec:
...
+                    area = w * h
+
+                    # 归一化特征
+                    cx, cy = cx / width, cy / height
+                    w, h = w / width, h / height
+                    area = area / (width * height)
+
+                    offset = object_id * 6
+                    feature[offset] += 1
+                    # TODO: 当前实现为直接取代旧值，并未完全处理同类多实例的情况
```
Comment (Jiadezhende) [inline app/services/inference/workflows/clean.py:212]: 仍然只保留一只手的信息吗

File: app/services/inference/workflows/clean.py [hunk 8]
```diff
@@ -96,3 +102,111 @@ def prepare_visualization_data(self, output: FrameDetections) -> RenderSpec:
...
+        if not aligned_frames:
+            return torch.tensor([])
+        
+        features = []
+
+        for aligned in aligned_frames:
+            feature = [0.0] * (self.num_objects * 6)
+
+            for frame in aligned.by_source.values():
+                width, height, _ = frame.metadata["frame_shape"]
```
Comment (Jiadezhende) [inline app/services/inference/workflows/clean.py]: 宽高搞反了，frame.shape 是 (H, W, C)

File: app/services/inference/workflows/clean.py [hunk 9]
```diff
@@ -96,3 +102,111 @@ def prepare_visualization_data(self, output: FrameDetections) -> RenderSpec:
...
+                if width <= 0 or height <= 0:
+                    continue
+                for detection in frame.detections:
+                    object_id = self._object_id(detection.class_name)
+                    # 忽略未检测到的物体
+                    if detection.bbox is None:
```
Comment (Jiadezhende) [inline app/services/inference/workflows/clean.py:189]: 未知类别这里返回了-1 → 特征列表被撑坏 / 崩溃

**Review by Jiadezhende — CHANGES_REQUESTED** [review]

File: app/services/inference/temporal/operator.py [hunk 10]
```diff
@@ -91,3 +92,77 @@ def _zip_by_ts(
...
+
+        self.objects = objects  # id -> name
+        self.actions = actions  # id -> name
+
+        import torch
+        self._model: torch.nn.Module = None
+        self._model_load_lock = threading.Lock()
+        self._device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    def _object_id(self, object_name: str) -> int:
```
Comment (Jiadezhende) [inline app/services/inference/temporal/operator.py:125]: _object_id/_action_id 每次调用都重建反查字典，而它在"每帧每检测框"的热路径上被反复调用，反查表不变，应在 __init__ 里建一次

**Review by Jiadezhende — CHANGES_REQUESTED** [review]

File: app/services/inference/workflows/clean.py (hunk 7 above)
Comment (fenglinglingzi) [reply app/services/inference/workflows/clean.py:212]: 后续根据训练情况，调整特征选择

**Review by Jiadezhende — APPROVED** [review]
